# Exploring eBay Car Data
---------------------

### <a id = 0> Table of contents:</a>

1. [Section 1: Introduction](#1)
2. [Section 2: Cleaning Column Names](#2)
3. [Section 3: Initial Exploration and Cleaning](#3)
4. [Section 4: Exploring the Odometer and Price Columns](#4)
5. [Section 5: Exploring the date columns](#5)
6. [Section 6: Dealing with incorrect Registration year Date](#6)
7. [Section 7: Exploring Price by Brand](#7)
8. [Section 8: Storing Aggregated Data in a DataFrame](#8)
9. [Section 9: Next steps](#9)
    - Data Cleaning:
        - [Translate and explore categorical data](#911)
        - [Converting dates into uniform numeric data](#912)
    - Analysis next step:
        - [Find the most common brand/model combinations](#921)
        - [Inspect a possible relation between price and milage](#922)
        - [Compare damaged and non-damaged cars average prices](#923)

In [1]:
# Imports packages
import numpy as np
import pandas as pd

In [2]:
# Uses detect method from the 'chardet' package to find file encoding
def encode(filepath):
# Reads the file as bytes
    import chardet
    f = open(filepath, 'rb')
    data = f.read()
    f.close()
# Detects the encoding
    encoding = chardet.detect(data)['encoding']
    return encoding

# Reads data into DataFrame
autos = pd.read_csv('autos.csv', encoding = encode('autos.csv'))

In [3]:
# Prints overview of DataFrame
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
# Explores autos DataFrame, columens and data typs
print(autos.info())
print(autos.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

#### <a id = 2>Section 2: Cleaning Column Names</a> - [Top](#0)

__What I see in <span style ='color:blue'> _"autos"_ </span> dataframe:__
----

- The dataframe has 50000 rows with 20 columns including five integers and 15 objects (string). There are Date and Numeric data imported as objects which may need data type conversion. 
- There are five columns with null values. However, the number of null values are less than 10000 for each of these columens.
- Column labels may need to be modified as some are long, and some have uppercase letters.
- There are Date columns that need data type conversion.
- Some columns with potentially numeric values have string type as measurement unit was included as a character.

#### Column Labels

Python's prefered convention for column name is [Snakecase](https://en.wikipedia.org/wiki/Snake_case) while in autos dataframe, the [Camelcase](https://en.wikipedia.org/wiki/Camel_case) convention was applied.
The Camelcase convention can be confusing, as it uses uppercases to separate words when I am writing codes. I have done the following changes to column names at this step. More adjustment may apply at further steps to make them more informatic, e.g. measurement units.

|Column name|Modified Column name|
|-----------|--------------------|
|dateCrawled|ad_crawled| 
|offerType|offer_type|
|vehicleType|vehicle_type|
|yearOfRegistration|registration_year|
|monthOfRegistration|registration_month|
|fuelType|fuel_type|
|notRepairedDamage|unrepaired_damage|
|dateCreated|ad_created|
|nrOfPictures|picture_counts|
|postalCode|postcode|
|lastSeen|last_seen|

In [5]:
# Renames columns by array edit

rename_dict = {'dateCrawled':'date_crawled',
               'offerType':'offer_type',
                'vehicleType':'vehicle_type',
                'yearOfRegistration':'registration_year',
                'powerPS':'power_ps',
                'monthOfRegistration':'registration_month',
                'fuelType':'fuel_type',
                'notRepairedDamage':'unrepaired_damage',
                'dateCreated':'ad_created',
                'nrOfPictures':'picture_counts',
                'postalCode':'postcode',
                'lastSeen':'last_seen'
              }
col_names = autos.columns
new_col_names =[]
for col in col_names:
    if col in rename_dict:
        col = rename_dict[col]
    new_col_names.append(col)

autos.columns = new_col_names
   
print(autos.columns)
print(autos.head()) 

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'picture_counts', 'postcode',
       'last_seen'],
      dtype='object')
          date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine             

In [6]:
# Displays descriptive statistics of all columns
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 19:38:20  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  picture_counts      postcode            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [7]:
# Checks picture_count column values
print(autos['picture_counts'].value_counts()) 

# Investigates price column
print(autos['price'].value_counts()) 

# Explores registration_year column values
print(autos['registration_year'].value_counts())

0    50000
Name: picture_counts, dtype: int64
$0         1421
$500        781
$1,500      734
$2,500      643
$1,000      639
           ... 
$2,479        1
$9,655        1
$4,210        1
$21,790       1
$6,998        1
Name: price, Length: 2357, dtype: int64
2000    3354
2005    3015
1999    3000
2004    2737
2003    2727
        ... 
1931       1
1929       1
1001       1
9996       1
1952       1
Name: registration_year, Length: 97, dtype: int64


#### <a id = 3>Section 3: Initial Exploration and Cleaning</a> -[Top](#0)

--------------------
Before any statistical analysis or even visualization of data, I need to prepare them by cleaning, including the investigation of outliers, finding missing data, data type conversion and so on. Here are what I noticed and what interventions need to be applied based on the result of my primary investigation:

__Columns with no informative values:__ 

There are three columns with nearly one value which do not add any information to my analysis. I am removing the following columns:
> - `seller`: contains one unique value for 49999 records.
> - `offer_type`: same as seller column, 49999 values are the same.
> - `picture_counts`: all values are 0.

__Columns that may need datatype conversion:__ 

The following columns have the potential to change from a string datatype into numeric or DateTime datatype. However, in some cases, data cleaning may be necessary to delete non-digits characters from the string before conversion.
- `ad_crawled`, `last_seen`: the values are recorded as a string and have a patter of `"yyyy-mm-dd hh:mm:ss"` with 24 hours format. I may convert this column's datatype to DateTime.
- `price`: datatype can be numeric. However, I need to remove the dolour sign\(\$\) before datatype conversion.
- `abtest`: there are two unique values which indicate that I can change its datatype into a categorical data type.
- `vehicle_type`: this column contains 7 unique values with around 10% null value. I am going to change its data type to categorical.
- `gearbox`: The two unique values in this column indicates that I can change its datatype to a category.
- `odometer`: contains digits with added 'km' characters indicating the unit of measurement. I will add the 'km' to column name after converting to integer datatype indicating the unit of measurement. 

__Columns may need more further investigation:__ 

- `registration_year`: A numeric column with odd maximum (9999) and minimum(1000) values for the year. This column needs further investigation concerning outliers.

__Dealing with null values:__ 

- `price`: this column has 1421 records with an amount equal to 0. This value may represent the null value. I may need to change 0 values into NaN as the 0 may affect the real descriptive statistics analysis figures. 
- `power_ps`: The minimum value of 0 may indicate the null value and conversion to NaN may be necessary to achieve correct descriptive stats.

To follow project instruction, I will remove the non-numeric characters from the `price` and `odometer` columns and cornvert them to a numeric dtype. The non-numeric characters include 'km' for `odometer`, dolour sign '$' for `price` and thousand separator comma (',') for both columns.

In [8]:
# Removes uninformative columns
autos = autos.drop(columns = ['seller', 'offer_type', 'picture_counts'])
print(autos.shape)

(50000, 17)


In [9]:
# Price column, cleans and converts dtype to int 
autos['price'] = autos['price'].str.replace('$', '')
autos['price'] = autos['price'].str.replace(',', '')
autos['price'] = autos['price'].astype(int)

# Odometer, clean and convert type to int
autos['odometer'] = autos['odometer'].str.replace('km', '')
autos['odometer'] = autos['odometer'].str.replace(',', '')
autos['odometer'] = autos['odometer'].astype(int)

# Renames column name
autos.rename(columns ={'odometer':'odometer_km'}, inplace = True)
print(autos['price'].dtype)
print(autos['odometer_km'].dtype)
print(autos.loc[:,['price', 'odometer_km']].describe())

int64
int64
              price    odometer_km
count  5.000000e+04   50000.000000
mean   9.840044e+03  125732.700000
std    4.811044e+05   40042.211706
min    0.000000e+00    5000.000000
25%    1.100000e+03  125000.000000
50%    2.950000e+03  150000.000000
75%    7.200000e+03  150000.000000
max    1.000000e+08  150000.000000


#### <a id = 4> Section 4: Exploring the Odometer and Price Columns</a> - [Top](#0)
At this part, I explored outlier values in `odometer_m` and `price` columns that I have worked on the previous section.

#### `price`: 

It is challenging to determine outlier without any information about the valid price range. For instances, there are really low prices but challenging to decide as outliers; although, the values are not realistic.
However, based on limited information and following the discussion section of the project, I will clean outliers based on the distribution of the values and some assumptions of the realistic price range:

In [10]:
# Defines a function for exploring a data range
def describe_range(df, col, l = 0, h = 100000000):
    df_sub = df.loc[df[col].between(l,h), col]
    print('Number of Unique values in {} and {}:'.format(l,h), df_sub.unique().shape[0])
    print('Number of records in {} and {}:'.format(l,h), df_sub.shape[0])
    print(df_sub.value_counts())
    print(df_sub.value_counts().sort_index())

# Displays descriptive analysis of price column
print(autos['price'].unique().shape[0])
print(autos['price'].describe())
print(autos['price'].value_counts().sort_index(ascending = False).head())
print(autos['price'].value_counts().sort_index().head())

2357
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
Name: price, dtype: int64
0    1421
1     156
2       3
3       1
5       2
Name: price, dtype: int64


There are 2357 unique values from the minimum figure of '0' to maximum '99999999'(rounded up to 1.0e+8). One approach to identify the outliers range is using the Interquartile range(IQR).

In [11]:
# Calculates the upper fence for strong outliers
iqr = 7200 - 1100
upper_rng = 7200 + (iqr * 3)
describe_range(autos, 'price', l = upper_rng)

Number of Unique values in 25500 and 100000000: 485
Number of records in 25500 and 100000000: 1373
26500     30
27500     27
25500     26
26900     24
28000     23
          ..
39979      1
194000     1
58900      1
72500      1
63499      1
Name: price, Length: 485, dtype: int64
25500       26
25590        1
25600        2
25650        2
25698        1
            ..
10000000     1
11111111     2
12345678     3
27322222     1
99999999     1
Name: price, Length: 485, dtype: int64


It appears that most of the data are concentrated below 25500. There are 1373 outliers higher than 25500. I adjusted the upper range into 36000, following further inspection to keep more data.

In [12]:
# Explores adjusted upper cutoff point of 36000
describe_range(autos, 'price', l = 36001)

Number of Unique values in 36001 and 100000000: 270
Number of records in 36001 and 100000000: 480
39900     12
37900      9
39500      9
38500      8
38900      8
          ..
67911      1
36675      1
104900     1
46911      1
128000     1
Name: price, Length: 270, dtype: int64
36500       4
36675       1
36800       1
36850       1
36900       4
           ..
10000000    1
11111111    2
12345678    3
27322222    1
99999999    1
Name: price, Length: 270, dtype: int64


There are 480 outliers above the new adjusted upper cutoff point with 270 unique values which indicate the average of fewer than two records per figure. 

In [13]:
# Explores the lower cutoff point and how many outliers are there
lower_rng = 99
describe_range(autos, 'price', h = lower_rng)

Number of Unique values in 0 and 99: 37
Number of records in 0 and 99: 1762
0     1421
1      156
50      49
99      19
80      15
70      10
60       9
30       7
10       7
40       6
75       5
90       5
25       5
65       5
49       4
45       4
20       4
2        3
17       3
12       3
55       2
5        2
15       2
13       2
11       2
9        1
89       1
3        1
8        1
14       1
66       1
47       1
18       1
29       1
59       1
35       1
79       1
Name: price, dtype: int64
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
40       6
45       4
47       1
49       4
50      49
55       2
59       1
60       9
65       5
66       1
70      10
75       5
79       1
80      15
89       1
90       5
99      19
Name: price, dtype: int64


The lower price range for outliers cannot be calculated by IQR method as the result of %25 percentile minus three times IQR falls at a negative figure. However, I assume that the prices lower than 100, are not reasonable. At this cutoff point, 1762 record would be removed which 1421 of those contain 0 value.

In [14]:
# Explores price column after outlier removed
autos_tmp = autos.copy() # initiates a temporary copy of dataset to work with
autos_tmp = autos_tmp[autos_tmp['price'].between(100, 36000)]
print(autos_tmp['price'].shape[0])
print(autos_tmp['price'].unique().shape[0])
print(autos_tmp['price'].describe())
print(autos_tmp['price'].value_counts())
print(autos_tmp['price'].value_counts().sort_index())

47758
2050
count    47758.000000
mean      5396.739457
std       6117.945858
min        100.000000
25%       1250.000000
50%       3000.000000
75%       7250.000000
max      36000.000000
Name: price, dtype: float64
500      781
1500     734
2500     643
1200     639
1000     639
        ... 
6410       1
5049       1
30950      1
3318       1
1775       1
Name: price, Length: 2050, dtype: int64
100      134
110        3
111        2
115        2
117        1
        ... 
35950      1
35980      1
35990      1
35999      2
36000     12
Name: price, Length: 2050, dtype: int64


In [15]:
# Finalizes changes by updating original dataframe
autos = autos_tmp

Prices lower than 100 and greater than 36,000 were removed as outliers. The total number of data reduced to 47,758 from 50,000. Exploring the distribution of data shows that 50 percent of cars priced between 1,250 to 7,250 dollars and the cheapest and most expensive vehicles are 100 and 36,000 dollars as expected.

### `odometer`: 

In [16]:
# Explores odometer
print(autos['odometer_km'].unique().shape)
print(autos['odometer_km'].describe())
print(autos['odometer_km'].value_counts())
print(autos['odometer_km'].value_counts().sort_index())

(13,)
count     47758.000000
mean     126586.017002
std       38925.687202
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
150000    31174
125000     5008
100000     2075
90000      1712
80000      1392
70000      1189
60000      1110
50000       971
40000       766
30000       727
5000        727
20000       692
10000       215
Name: odometer_km, dtype: int64
5000        727
10000       215
20000       692
30000       727
40000       766
50000       971
60000      1110
70000      1189
80000      1392
90000      1712
100000     2075
125000     5008
150000    31174
Name: odometer_km, dtype: int64


The descriptive analysis of `odometer_km` column shows that there is no missing value, and we have 13 unique values in this column from minimum 5,000 up to maximum 150,000 km. Limited unique values indicate that the website provides a list to select odometer when ad registered. 
More than 60 percent of values are distributed at 150,000 km with 31,174 instances. This may be due to data entry rules at the website which accepts the maximum value of 150,000 km even for higher figures.
This column does not seem to have any outlier, and I keep them as it is.

#### <a id = 5>Section 5: Exploring the date columns</a> - [top](#0)

There are five columns which represent date and their values provided from two different sources, crawler and website itself.

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

The first three columns are imported as string while the last two have `int64` data type.
At this section, I will prepare the first three columns for descriptive analysis and explore their distribution. 

In [17]:
# Searches for a pattern in cols_dt
cols_dt = ['date_crawled', 'ad_created', 'last_seen']
print(autos[cols_dt].head())

          date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50


In [18]:
# Extracts the first 10 characters of each value for all three columns
# as they have the same pattern and I am interested in part representing date
autos_tmp = autos.copy() # Creates temporary copy of dataframe to apply changes
for col in cols_dt:
    autos_tmp[col] = autos_tmp[col].str[:10]
print(autos_tmp[cols_dt].head())

  date_crawled  ad_created   last_seen
0   2016-03-26  2016-03-26  2016-04-06
1   2016-04-04  2016-04-04  2016-04-06
2   2016-03-26  2016-03-26  2016-04-06
3   2016-03-12  2016-03-12  2016-03-15
4   2016-04-01  2016-04-01  2016-04-01


In [19]:
# Finalizes changes by updating dataframe
autos = autos_tmp

In [20]:
# Defines a function for inspecting date columns
def describe_dt_cols(df, col, ascend = True):
    print(df[col].describe())
    print(df[col].value_counts(dropna = False))
    print(df[col].value_counts(normalize = True, dropna = False)
                  .sort_index(ascending = ascend))
    
# Explors date_crawled column
describe_dt_cols(autos, 'date_crawled', ascend = True)

count          47758
unique            34
top       2016-04-03
freq            1849
Name: date_crawled, dtype: object
2016-04-03    1849
2016-03-20    1806
2016-03-12    1770
2016-03-21    1766
2016-03-14    1753
2016-04-04    1749
2016-03-07    1725
2016-04-02    1698
2016-03-28    1665
2016-03-19    1658
2016-03-15    1639
2016-03-29    1633
2016-03-30    1615
2016-04-01    1608
2016-03-08    1588
2016-03-09    1580
2016-03-22    1568
2016-03-11    1551
2016-03-26    1550
2016-03-10    1543
2016-03-23    1541
2016-03-31    1518
2016-03-17    1503
2016-03-25    1500
2016-03-27    1485
2016-03-16    1408
2016-03-24    1405
2016-03-05    1214
2016-03-13     749
2016-03-06     671
2016-04-05     617
2016-03-18     614
2016-04-06     152
2016-04-07      67
Name: date_crawled, dtype: int64
2016-03-05    0.025420
2016-03-06    0.014050
2016-03-07    0.036120
2016-03-08    0.033251
2016-03-09    0.033083
2016-03-10    0.032309
2016-03-11    0.032476
2016-03-12    0.037062
2016-03-13    0.015

#### `date_crawled` column's exploration note:

- The original `date_crawled` column contains the date and time of the add crawled. I restricted the information it provides to the date only, and time part was excluded. Consequently, there are 34 unique date values which show that `autos` dataset were polled records with crawled date from 5-March-2016 till 7-April-2016.
- There is no missing or outlier value in this column.
- The majority of data distributed evenly over most days by around 3% of total data. However, there are two distinctive drops at 6th and 7th of April with less than 1 percent (minimum 71 records at 7/4/2016) of total data. These low figures may have happened because the last days have not been updated at the time data where collected.
- Some moderate downward fluctuation was noticed at 6th, 13th and 18th of March with 5 and 7 days period which may be related to weekly Trend or source update schedule or market trend.
- The maximum number of adds crawled at 3rd April 2016 with 1845 records.

In [21]:
# Exploring ad_created column
describe_dt_cols(autos, 'ad_created', ascend = True)
print(autos['ad_created']
      .value_counts(normalize =True, dropna = False)
      .describe())
print(autos['ad_created']
      .value_counts()
      .sort_index(ascending=True).head(35))
print(autos['ad_created']
      .value_counts()
      .sort_index(ascending=True).tail(5))
print(autos[autos['ad_created'].between('2015-08-10', '2016-02-26')].shape[0])
print(autos.loc[autos['ad_created'].between('2015-08-10', '2016-02-26'), 'ad_created'].unique().shape[0])

count          47758
unique            75
top       2016-04-03
freq            1860
Name: ad_created, dtype: object
2016-04-03    1860
2016-03-20    1809
2016-03-21    1775
2016-04-04    1767
2016-03-12    1762
              ... 
2016-01-13       1
2016-02-11       1
2016-02-16       1
2015-08-10       1
2016-01-29       1
Name: ad_created, Length: 75, dtype: int64
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
                ...   
2016-04-03    0.038946
2016-04-04    0.036999
2016-04-05    0.011642
2016-04-06    0.003266
2016-04-07    0.001256
Name: ad_created, Length: 75, dtype: float64
count    75.000000
mean      0.013333
std       0.015991
min       0.000021
25%       0.000021
50%       0.000168
75%       0.032309
max       0.038946
Name: ad_created, dtype: float64
2015-08-10    1
2015-09-09    1
2015-11-10    1
2015-12-05    1
2015-12-30    1
2016-01-03    1
2016-01-07    1
2016-01-10    2
2016-01-13    1
2016-

#### `ad_created` column's exploration note:

There are 47,758 adds created from "10-08-2015" up to "7-04-2016" with no null values. the following observations were noted after descriptive analysis:
- There are 75 unique dates that new ads created. Further investigation shows that there are only 54 new ads from "10-08-2015" to "26-02-2016".
- The maximum of 1946 new adds created on "03-04-2016".
- There is a surge drop in the number of new adds at 6th and 7th of April 2016.

In [22]:
# Explores last_seen column
describe_dt_cols(autos, 'last_seen', ascend = True)
print(autos['last_seen']
      .value_counts(normalize =True, dropna = False)
      .describe())
print(autos['last_seen']
      .value_counts()
      .sort_index(ascending=True).tail(5))

count          47758
unique            34
top       2016-04-06
freq           10577
Name: last_seen, dtype: object
2016-04-06    10577
2016-04-07     6273
2016-04-05     5940
2016-03-17     1346
2016-04-03     1206
2016-04-02     1192
2016-03-30     1185
2016-04-04     1180
2016-03-12     1142
2016-03-31     1139
2016-04-01     1097
2016-03-29     1072
2016-03-22     1024
2016-03-28     1001
2016-03-20      986
2016-03-21      982
2016-03-24      946
2016-03-25      914
2016-03-23      890
2016-03-26      797
2016-03-16      785
2016-03-15      762
2016-03-19      758
2016-03-27      744
2016-03-14      604
2016-03-11      596
2016-03-10      512
2016-03-09      460
2016-03-13      426
2016-03-08      353
2016-03-18      350
2016-03-07      259
2016-03-06      208
2016-03-05       52
Name: last_seen, dtype: int64
2016-03-05    0.001089
2016-03-06    0.004355
2016-03-07    0.005423
2016-03-08    0.007391
2016-03-09    0.009632
2016-03-10    0.010721
2016-03-11    0.012480
2016-03-12    

#### `last_seen` column's exploration note:

This column keeps the last date that an add viewed. A crawler adds this date just like `date_crawled` column. 
I noted the following observations after descriptive analysis:
- The recorded period is from 5-03-2016 up to 7-04-2016 with 34 unique dates, with no missing value. This period confirms that the dates follow the same pattern as carawled_date.
- 22% of adds (10,577 records) were last seen at 6th April 2016 and the number of instances of each day increases gradually from March to April which can be explained by the pattern of the number of adds created.
- There is some upward fluctuation on 12th, 17th and 22nd 0f March which may indicate a weekly trend of the market.

In [23]:
# Explores registration_year column

print('series length', autos['registration_year'].shape[0])
print(autos['registration_year'].describe())
print('Number of Unique values', autos['registration_year'].unique().shape[0])
# sorted unique values - ascending
print(np.sort(autos['registration_year'].unique()))
print('Top 5 value counts sorted by frequency ')
print(autos['registration_year']
      .value_counts(dropna=False)
      .head(5)
     )

series length 47758
count    47758.000000
mean      2004.614996
std         86.204350
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64
Number of Unique values 92
[1000 1001 1111 1800 1910 1927 1929 1931 1934 1937 1938 1939 1941 1948
 1950 1951 1952 1953 1954 1956 1957 1958 1959 1960 1961 1962 1963 1964
 1965 1966 1967 1968 1969 1970 1971 1972 1973 1974 1975 1976 1977 1978
 1979 1980 1981 1982 1983 1984 1985 1986 1987 1988 1989 1990 1991 1992
 1993 1994 1995 1996 1997 1998 1999 2000 2001 2002 2003 2004 2005 2006
 2007 2008 2009 2010 2011 2012 2013 2014 2015 2016 2017 2018 2019 2800
 4100 4500 4800 5000 5911 8888 9000 9999]
Top 5 value counts sorted by frequency 
2000    3102
2005    2907
1999    2878
2003    2692
2004    2691
Name: registration_year, dtype: int64


#### `registration_year` column's exploration note:

This column contains the vehicle's year of registration with 47,758 records and no missing value. However, there are some outliers and unrealistic values which need attention as follows:

- There are 92 unique values in this column with a minimum of 1000 up to 9999. The min and max values are unrealistic figures for the year, which urges more investigation to identify odd values.
- The maximum number of cars for sale (3102 adds) have registered in 2000, nearly 16 yeas old vehicles.

In [24]:
# Defines and removes outliers
print(autos['registration_year']
      .value_counts(dropna=False)
      .sort_index()
      .head(30)
     )
print(autos['registration_year']
      .value_counts(dropna=False)
      .sort_index()
      .tail(10)
     )

# Identifies the number of outliers outside 1900-2016
print('number of outliers outside 1900-2016:', 
      autos[~(autos['registration_year'].between(1900,2016))].shape[0])
# Identifies the number of outliers over 2016
print('number of outliers over 2016:', 
      autos[autos['registration_year'].between(2017,9999)].shape[0])
# Identifies the number of outliers less than 1981
print('number of outliers less than 1981:', 
      autos[autos['registration_year'].between(1000,1980)].shape[0])
# Identifies the number of outliers less than 1960
print('number of outliers less than 1960:', 
      autos[autos['registration_year'].between(1000,1959)].shape[0])
# Identifies the number of outliers outside 1960-2016
print('number of outliers outside 1960-2016:', 
      autos[~(autos['registration_year'].between(1960,2016))].shape[0])

1000     1
1001     1
1111     1
1800     2
1910     2
1927     1
1929     1
1931     1
1934     2
1937     4
1938     1
1939     1
1941     2
1948     1
1950     1
1951     1
1952     1
1953     1
1954     2
1956     4
1957     2
1958     3
1959     6
1960    21
1961     5
1962     4
1963     7
1964    11
1965    16
1966    19
Name: registration_year, dtype: int64
2019    2
2800    1
4100    1
4500    1
4800    1
5000    3
5911    1
8888    1
9000    1
9999    3
Name: registration_year, dtype: int64
number of outliers outside 1900-2016: 1865
number of outliers over 2016: 1860
number of outliers less than 1981: 541
number of outliers less than 1960: 42
number of outliers outside 1960-2016: 1902


#### <a id = 6>Section 6: Dealing with incorrect Registration year Date</a> ([Top](#0))

#### Define outlier boundaries and remove them from `year_registration`

- The vehicle's registration date cannot be later than the date it listed. From the `ad_created` records, we know that the last adds created at 07-04-2016. Thus, the highest acceptable year value in `year_registration` is 2016. However, The month of registration plays a crucial role, as part of the year 2016 is acceptable while cars registered after April fall outside of the upper fence of acceptable values. There are 1860 cars registered after 2016.
- The Inter Quartile Range (IQR) is from 1999 to 2008. Then, based on the IQR approach to identify outliers, The cars registered before 1981  assumed outliers (541 outliers). But, with further investigation, I decided to adjust this lower cutoff point into 1960, which consequently drops outliers to 42.
- There are 1865 cars registered outside of 1900-2016 interval, while there are 1902 outliers with 1960-2016 interval.

In [25]:
##### Removes outliers based on 1960-2016 interval of year_registration
## Initiates temporary copy
autos_tmp = autos.copy()
autos_tmp = autos_tmp[autos_tmp['registration_year'].between(1960,2016)]

# Explores the distribution of year_registration following outliers removal
print(autos_tmp['registration_year'].describe())
print(autos_tmp['registration_year'].unique().shape)
print(autos_tmp['registration_year']
      .value_counts(dropna=False, normalize=True)
     )
print(autos_tmp['registration_year']
      .value_counts(dropna=False, normalize=True)
      .sort_index()
     )

count    45856.000000
mean      2002.940008
std          6.843462
min       1960.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64
(57,)
2000    0.067647
2005    0.063394
1999    0.062762
2003    0.058706
2004    0.058684
2006    0.058095
2001    0.057288
2002    0.053930
1998    0.050942
2007    0.049219
2008    0.047736
2009    0.045120
1997    0.041892
2011    0.034630
2010    0.034107
1996    0.029505
2012    0.027128
1995    0.025951
2016    0.025907
2013    0.016268
1994    0.013542
2014    0.013128
1993    0.009094
1992    0.007938
1991    0.007327
1990    0.007153
2015    0.006673
1989    0.003729
1988    0.002879
1985    0.002006
1980    0.001745
1987    0.001548
1986    0.001527
1984    0.001112
1983    0.001090
1982    0.000894
1978    0.000894
1970    0.000807
1979    0.000741
1972    0.000720
1981    0.000567
1968    0.000523
1974    0.000523
1967    0.000502
1973    0.000502
1971    0.000

In [26]:
# Finalizes changes to dataframe
autos = autos_tmp

#### Descriptive analysis of `year_registration` after removing outliers

- The final dataframe contains 45,856 records with the unique values dropped from 92 to 57 for the year of registration.
- The distribution of data did not change as the majority of data were already distributed at the same interval. The IQR and 50% percentile remained the same as 1999-2008 and 2003 respectively.
- As expected, the average registration year dropped from 2004.61 to 2002.94. The distribution of removed outliers higher (1860 outliers) and lower (42 outliers) cut-off points explains the reason.

#### <a id = 7>Section 7: Exploring Price by Brand</a> - [Top](#0)

------------------

Here, I use aggregation to explore the average price of different car brands.

In [27]:
# Explores brand column
print('The number of car brands listed:', autos['brand'].unique().shape[0])
print(autos['brand'].value_counts(normalize=True, dropna=False))

The number of car brands listed: 40
volkswagen        0.213036
bmw               0.110040
opel              0.108317
mercedes_benz     0.095800
audi              0.085725
ford              0.070285
renault           0.047540
peugeot           0.030181
fiat              0.025885
seat              0.018449
skoda             0.016574
nissan            0.015505
mazda             0.015374
smart             0.014349
citroen           0.014175
toyota            0.012910
hyundai           0.010119
volvo             0.009225
mini              0.008897
sonstige_autos    0.008636
mitsubishi        0.008243
honda             0.007938
kia               0.007153
alfa_romeo        0.006695
suzuki            0.005997
chevrolet         0.005670
chrysler          0.003555
porsche           0.003358
dacia             0.002682
daihatsu          0.002530
jeep              0.002181
subaru            0.002137
land_rover        0.001766
saab              0.001679
jaguar            0.001505
daewoo            0

There are 40 brands of cars listed for sale. Exploring relative frequency shows more than 52% of listed vehicles are from 5 German Automotive companies as is expected from German eBay car listing website. By contrast, the Lada (a Russian brand) is the least popular car.
I confined my investigation to the brands with a relative frequency of more than 1%.

In [28]:
# Selects brands with relative frequencies greater than 1%
brands = autos['brand'].value_counts(normalize=True, dropna=False)
brands = brands[brands > 0.01].index.tolist()
brands_price_avg = {}
print("Sorted by brand's popularity")
# Calculates the average price of selected brands
for brand in brands:
    price_avg = (autos.loc[autos['brand'] == brand, 'price']
                               .mean()
          )
    brands_price_avg[brand] = round(price_avg)
    print('{} --> {:.0f}'.format(brand,round(price_avg)))
# Sorts brands by the average price (descending)
sorted_brands_price_avg = sorted(brands_price_avg.items(), key = lambda x: x[1], reverse = True)
print('\n')
print('Sorted by average price')
for i in sorted_brands_price_avg:
    print('{} --> {:.0f}'.format(i[0],i[1]))

Sorted by brand's popularity
volkswagen --> 5324
bmw --> 7832
opel --> 2997
mercedes_benz --> 7830
audi --> 8495
ford --> 3596
renault --> 2455
peugeot --> 3114
fiat --> 2837
seat --> 4433
skoda --> 6417
nissan --> 4757
mazda --> 4080
smart --> 3596
citroen --> 3779
toyota --> 5092
hyundai --> 5411


Sorted by average price
audi --> 8495
bmw --> 7832
mercedes_benz --> 7830
skoda --> 6417
hyundai --> 5411
volkswagen --> 5324
toyota --> 5092
nissan --> 4757
seat --> 4433
mazda --> 4080
citroen --> 3779
ford --> 3596
smart --> 3596
peugeot --> 3114
opel --> 2997
fiat --> 2837
renault --> 2455


- The first top three brands are from German Automotive companies. Audi has the highest average price of 8495 \$, followed by BMW and Mercedes Benz.
- Renault has the lowest average price in my selected brands by 2455 Dollars.
- The higher popularity of Volkswagon can be explained by its lower average price \(5324 \$\) and being a domestic manufactured vehicle.
- Asian popular companies like Toyota, Nissan, and Hyundai are more expensive on average compare to Europan car, like Citroen, Peugeot and Renault.
- The popularity of brands do not merely depend on their average price. Some domestic or European brands are more popular than East Asian brands. However, within each origin group, brands with lower average price are mostly more popular. 

#### <a id= 8>Section 8: Storing Aggregated Data in a DataFrame:</a> ([Top](#0)) 

I am going to store the average price and average mileage (`odometer_km`)  of the top 6 listed brands in a dataframe.

In [29]:
# Creates a 'top six brands' index
top_6 = (autos['brand']
             .value_counts(normalize=True)
             .sort_values(ascending = False)
             .head(6)
        )
brands_top_6 = top_6.index

# Calculates the mean milage as brand mean mileage (bmm) dictionary
bmm_dic = {}
for brand in brands_top_6:
    milage_avg = (autos.loc[autos['brand'] == brand, 'odometer_km']
                     .mean())
    bmm_dic[brand] = round(milage_avg)

# Calculates the mean price as brand mean price (bmp) dictionary
bmp_dic = {}
for brand in brands_top_6:
    price_avg = (autos.loc[autos['brand'] == brand, 'price']
                     .mean())
    bmp_dic[brand] = round(price_avg)

# Constructs pandas series of average price and mileage
bmp_series = pd.Series(bmp_dic) # brands mean price series
bmm_series = pd.Series(bmm_dic) # brands mean milage series

# Stores series in a DataFrame
brands_df = pd.DataFrame(bmp_series, columns = ['mean_price'])
brands_df['mean_milage'] = bmm_series
brands_df['frequency'] = top_6

# Sorts data by mean price value (descending)
print('sorted by Price:')
print(brands_df.sort_values('mean_price', ascending = False))
print('\n')
print('sorted by milage:')
print(brands_df.sort_values('mean_milage', ascending = False))

sorted by Price:
               mean_price  mean_milage  frequency
audi               8495.0     131235.0   0.085725
bmw                7832.0     133703.0   0.110040
mercedes_benz      7830.0     132407.0   0.095800
volkswagen         5324.0     129038.0   0.213036
ford               3596.0     124592.0   0.070285
opel               2997.0     129412.0   0.108317


sorted by milage:
               mean_price  mean_milage  frequency
bmw                7832.0     133703.0   0.110040
mercedes_benz      7830.0     132407.0   0.095800
audi               8495.0     131235.0   0.085725
opel               2997.0     129412.0   0.108317
volkswagen         5324.0     129038.0   0.213036
ford               3596.0     124592.0   0.070285


A few findings from the above:
- No specific relationship between the average price and average mileage emerged for the top six cars.
- Audi has the highest price of 8495 \$ with third-highest average mileage of 131,235 km.
- Volkswagen's average price is higher than Ford and Opel, although the average mileage is higher than Ford and approximately equal to Opel. However, Volkswagen is more popular.

#### <a id = 9>Section 9: Next Steps</a> - [Top](#0)
--------------
Within this section, I am going to do further Data cleaning, and Analysis suggested in the last part of the Pandas Guided Project:

- Data cleaning next step:
    - To identify Categorical data, translate the values into English
    - To convert dates into uniform numeric data
- Analysis next step:
    - To find the most popular brand/model combination
    - To inspect possible relation between average price and milage by splitting `odometer_km` into groups and aggregating data.
    - To compare the average price of damaged cars with non-damaged

*__<a id = 911> Categorical columns Data cleaning:</a> ([Top](#0))__*

The `autos` dataframe has 5 columns with values that can be devided into groups(categories):

- `vehicle_type`
- `gearbox`
- `fuel_type`
- `unrepaired_damage`
- `abtest`

In [30]:
# Explores unique vales of categorical columns
for col in ['vehicle_type', 'gearbox', 'fuel_type', 
            'unrepaired_damage', 'abtest']:
    print(autos[col].unique())

['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere']
['manuell' 'automatik' nan]
['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere']
['nein' nan 'ja']
['control' 'test']


*__Substitutes English equivalent for vehicle_type column values using Series.map()__*

In [31]:
# Initiates temporary copy of dataframe
autos_tmp = autos.copy()
type_map = {'bus':'bus', 'limousine':'limousine', 'suv':'suv', 'coupe':'coupe', 
           'kleinwagen':'small_car', 'small_car':'small_car',
           'kombi':'station', 'station':'station', 
           'cabrio':'convertible', 'convertible':'convertible',
           'andere':'other', 'other':'other'}
autos_tmp['vehicle_type'] = autos_tmp['vehicle_type'].map(type_map)
autos_tmp = autos_tmp.astype({'vehicle_type':'category'})
print(autos_tmp['vehicle_type'].unique())
print(autos_tmp['vehicle_type'].dtype)
print(autos_tmp['vehicle_type'].value_counts(dropna=False))

[bus, limousine, small_car, station, NaN, coupe, suv, convertible, other]
Categories (8, object): [bus, limousine, small_car, station, coupe, suv, convertible, other]
category
limousine      12472
small_car      10496
station         8868
bus             3990
convertible     2887
NaN             2546
coupe           2338
suv             1879
other            380
Name: vehicle_type, dtype: int64


*__Gearbox column: changes dtype to category and replaces German words with English using category rename method__*

In [32]:
print(autos_tmp['gearbox'].value_counts(dropna=False))
autos_tmp['gearbox'] = autos_tmp['gearbox'].astype('category')
autos_tmp['gearbox'].cat.rename_categories({'automatik':'automatic'}, inplace=True)                   
print(autos_tmp['gearbox'].value_counts(dropna=False))
print(autos_tmp['gearbox'].dtype)

manuell      34398
automatik     9504
NaN           1954
Name: gearbox, dtype: int64
manuell      34398
automatic     9504
NaN           1954
Name: gearbox, dtype: int64
category


*__Fuel_type column: changes dtype to category and replaces german words with English using category rename method__*

In [33]:
print(autos_tmp['fuel_type'].value_counts(dropna=False))
autos_tmp['fuel_type'] = autos_tmp['fuel_type'].astype('category')
autos_tmp['fuel_type'].cat.rename_categories({'benzin':'petrol',
                                             'elektro':'electric',
                                             'andere':'other'}, inplace=True)                   
print(autos_tmp['fuel_type'].value_counts(dropna=False))
print(autos_tmp['fuel_type'].dtype)

benzin     28091
diesel     13834
NaN         3149
lpg          648
cng           70
hybrid        35
elektro       15
andere        14
Name: fuel_type, dtype: int64
petrol      28091
diesel      13834
NaN          3149
lpg           648
cng            70
hybrid         35
electric       15
other          14
Name: fuel_type, dtype: int64
category


*__unrepaired_damage column: translates to yes and no , changes dtype to category change dtype__*

In [34]:
autos_tmp['unrepaired_damage'] = autos_tmp['unrepaired_damage'].astype('category')
# rename ctegories value
autos_tmp['unrepaired_damage'].cat.rename_categories({'nein':'no', 'ja':'yes'}, inplace=True)
print(autos_tmp['unrepaired_damage'].value_counts(dropna=False))

no     33329
NaN     8057
yes     4470
Name: unrepaired_damage, dtype: int64


In [35]:
# Finalizes changes to dataframe
autos = autos_tmp

#### <a id = 912> Converting dates into uniform numeric data:</a> ([Top](#0))

There are 3 date string columns with "yyyy-mm-dd" pattern:
- `date_crawled`
- `ad_created`
- `last_seen`

These object columns will be converted into integer with uniform numeric value of 8 digits (eg, 20160924)

In [36]:
# Converts the date columns to uniform numeric data 
autos_tmp = autos.copy()
date_cols = ['date_crawled', 'ad_created', 'last_seen'] # date columns the will be converted

# Defines function to handle repetetive procedures
def date_convert(df, col):
    sr = df[col]
    sr = sr.str.replace('-', '') # remove non numeric characters
    sr = sr.astype(int) # change dtype
    return sr
# applies function on every date column
for col in date_cols:
    autos_tmp[col] = date_convert(autos_tmp, col)
    print(autos_tmp[col].head()) # check conversion

# Compares to see if changes occurred
print(autos.info())
print(autos_tmp.info())

0    20160326
1    20160404
2    20160326
3    20160312
4    20160401
Name: date_crawled, dtype: int64
0    20160326
1    20160404
2    20160326
3    20160312
4    20160401
Name: ad_created, dtype: int64
0    20160406
1    20160406
2    20160406
3    20160315
4    20160401
Name: last_seen, dtype: int64
<class 'pandas.core.frame.DataFrame'>
Int64Index: 45856 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   date_crawled        45856 non-null  object  
 1   name                45856 non-null  object  
 2   price               45856 non-null  int64   
 3   abtest              45856 non-null  object  
 4   vehicle_type        43310 non-null  category
 5   registration_year   45856 non-null  int64   
 6   gearbox             43902 non-null  category
 7   power_ps            45856 non-null  int64   
 8   model               43775 non-null  object  
 9   odometer_km         45856 non-nu

In [37]:
# Finalize changes to dataframe
autos = autos_tmp

#### <a id = 921> Find the most common brand/model combinations </a> [Top](#0) 

In [38]:
# Shortlists top 10 popular brands 
top_10 = autos['brand'].value_counts().head(10).index

# Initiates dictionaries to keep records 
bmpm_dic = {} # brand:most popular model
bmpmc_dic = {} # brand:most popular model

for brand in top_10:
    counts = autos.loc[autos['brand']==brand, 'model'].value_counts()
    bmpm_dic[brand] = counts.index[0]
    bmpmc_dic[brand] = counts[0]

bmpm_series = pd.Series(bmpm_dic)
bmpmc_series = pd.Series(bmpmc_dic)

# Constructs brand model counts dataframe
bmc_df = pd.DataFrame(bmpm_series, columns =['max_model']) 
bmc_df['max_counts'] = bmpmc_series

print('The most popular Brand/Model with its counts is:')
print(bmc_df.loc[bmc_df['max_counts']==bmc_df['max_counts'].max()])

# Alternative approuch printing The most popular model/brand with counts
# bmc_sorted = bmc_df.sort_values('max_counts', ascending=False)
# print(bmc_sorted.iloc[0,:])


The most popular Brand/Model with its counts is:
           max_model  max_counts
volkswagen      golf        3682


#### <a id = 922>Inspect a possible relation between price and mileage:</a> - [top](#0)

In [39]:
# Inspects the distribution of mileage to define the groups
print(autos['odometer_km'].value_counts())

# Splits mileage values into 6 groupsand record in a new column
autos['odometer_grps'] = pd.cut(x=autos['odometer_km'], 
                                    bins=[1, 20000, 40000, 70000, 100000, 150000, 200000],
                                    right=False)
# Inspects the new created column
print(autos['odometer_grps'].value_counts())

# Explores the average price for each odometer group
print(autos[['odometer_grps', 'price']].groupby(by=['odometer_grps']).mean())

# Prints average price without split into groups as there are limited unique values
print(autos[['odometer_km', 'price']].groupby(by=['odometer_km']).mean())

# Drops created odometer_grps column
autos.drop(columns = 'odometer_grps', inplace =True)
print(autos.shape)

150000    29846
125000     4809
100000     2015
90000      1649
80000      1351
70000      1158
60000      1080
50000       951
40000       747
30000       709
20000       672
5000        668
10000       201
Name: odometer_km, dtype: int64
[150000, 200000)    29846
[100000, 150000)     6824
[70000, 100000)      4158
[40000, 70000)       2778
[20000, 40000)       1381
[1, 20000)            869
Name: odometer_grps, dtype: int64
                         price
odometer_grps                 
[1, 20000)         7710.649022
[20000, 40000)    13726.301955
[40000, 70000)    11822.328654
[70000, 100000)    8986.998076
[100000, 150000)   6404.946073
[150000, 200000)   3722.577062
                    price
odometer_km              
5000          5347.064371
10000        15565.746269
20000        13897.415179
30000        13564.118477
40000        13050.152610
50000        12225.333333
60000        10618.215741
70000        10234.275475
80000         9182.937084
90000         7950.575500
100000    

The average price shows a pattern of decrease when odometer increases.
There is one exception at 5000 km odometer, which needs more investigation.

#### <a id = 923> Compare damaged and non-damaged cars average prices</a> - [Top](#0)

In [40]:
# Aggregates prices by unrrepaired_damage to find mean price
dudap =autos[['unrepaired_damage', 'price']].groupby(by=['unrepaired_damage']).mean()
print(dudap)

# Prints the result
print('The unrepaired damaged cars are {:.0%} cheaper.'.format(1 - dudap.loc['yes','price']/dudap.loc['no','price']))

                         price
unrepaired_damage             
yes                2207.283669
no                 6479.250803
The un_repaired damaged cars are 66% cheaper.
